### DISCUSSION TITLE
How to share DataTemplate s across projects?

### DISCUSSION THREAD

**Discussion by yll690** [AvaloniaUI/Avalonia#8426] [Q&A] [answered]
In WPF I can put the `DataTemplate`s in ResourceDictionary and merge it in other application. But in Avalonia there are many limits:

- `DataTemplate`s can't be put in ResourceDictionary without key. It's also not work when has a key like `vm:TestItem`.
- `DataTemplates` property is readonly.
- `DataTemplates` can not merge each other like `ResourceDictionary` or `Styles`.

So is there any way to share `DataTemplate`s across  projects?

**Comment by timunie** [comment, marked as answer]
If this is not yet possible it should be a feature request imo

For now you may create a helper class which implements IDataTemplate and handle the match there

**Reply by timunie** [reply]
Just a workaround for sure

**Reply by yll690** [reply]
Yes,  a feature request will be better. I just want to check if I missed something.
A custom `DataTemplateInclude` seems works for me. Thank you.


```
    public class DataTemplateInclude : IDataTemplate
    {
        private readonly Uri _baseUri;
        private DataTemplates _loaded;
        private bool _isLoading;

        public Uri Source { get; set; }

        public DataTemplates Loaded
        {
            get
            {
                if (_loaded == null)
                {
                    _isLoading = true;
                    _loaded = (DataTemplates)AvaloniaXamlLoader.Load(Source, _baseUri);
                    _isLoading = false;
                }

                return _loaded;
            }
        }


        public DataTemplateInclude(Uri baseUri)
        {
            _baseUri = baseUri;
        }

        public DataTemplateInclude(IServiceProvider serviceProvider)
        {
            _baseUri = serviceProvider.GetContextBaseUri();
        }


        public bool Match(object data)
        {
            if (_isLoading || Loaded == null)
                return false;

            return Loaded.Any(dt => dt.Match(data));
        }

        public IControl Build(object data)
        {
            if (_isLoading || Loaded == null)
                return null;

            return Loaded.FirstOrDefault(dt => dt.Match(data))?.Build(data);
        }
    }
```

**Reply by alexhelms** [reply]
This workaround worked for me and I just wanted to point our the external axaml file should look like this:

```
<DataTemplates xmlns="https://github.com/avaloniaui"
               xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml">

    <DataTemplate DataType="somewhere:Something">
        <TextBlock Text="{Binding}" />
    </DataTemplate>

</DataTemplates>
```

**Comment by maxkatz6** [comment]
> DataTemplates can't be put in ResourceDictionary without key. It's also not work when has a key like vm:TestItem.

It is possible to add DataTemplate to a resources and reference it with StaticResource when setting to a ContentTemplate or ItemTemplate. Same as in WPF.

> DataTemplates property is readonly.

It's a mutable collection. It should be readonly. Why would you need it writeable? For single template you should use ContentTemplate or ItemTemplate depending on a control.

> DataTemplates can not merge each other like ResourceDictionary or Styles.

True. But WPF also doesn't allow it from what I remember. You can use DataTemplateInclude as you created above, but it's not AOT safe.

> So is there any way to share DataTemplates across projects?

1. ResourceDictionary with DataTemplate objects inside. Reference them by a key.
2. Create XAML file with a class (using x:Class attribute), same would work in WPF from what I remember. After that you can reference DataTemplate with its class name `<local:MyDataTemplate />`.

**Reply by maxkatz6** [reply]
See https://github.com/AvaloniaUI/Avalonia/issues/2688 for details

**Reply by yll690** [reply]
Thank you for the detailed reply!

> and reference it with StaticResource when setting to a ContentTemplate or ItemTemplate.
> ResourceDictionary with DataTemplate objects inside. Reference them by a key.

Yes I can reference anything in `ResourceDictionary` with a key, but in Avalonia, `DataTemplate` can not apply to models automatically if it is in a `ResourceDictionary`. That's why `DataTemplates` property exsits, right? While in WPF, `DataTemplate` without a key (or with a Type key) can apply to models automatically. Now I also create a class to port `DataTemplate` from `ResourceDictionary` to `DataTemplates`.

> It's a mutable collection. It should be readonly

Reasonable. I was thinking that if it is writable, I can creat a `DataTemplates` in `ResourceDictionary` then assign it to a `DataTemplates` property, like the `ResourceDictionary` property. Maybe it's not a good idea. 

> True. But WPF also doesn't allow it from what I remember

No, there is no `DataTemplates`, no `Styles`, only `ResourceDictionary` in WPF, and the `ResourceDictionary` can merge.

> but it's not AOT safe.

I can't understand and maybe don't need it. However could you explain more? Does `StyleInclude` has the same problem?

Thanks again.

**Reply by maxkatz6** [reply]
> DataTemplate can not apply to models automatically if it is in a ResourceDictionary

It can. ```<ListBox ItemTemplate="{StaticResouce MyDataTemplate}" />```, where MyDataTemplate should have either empty DataType (match everything) or specific DataType (exact same as the model's type). ResourceDictionary can merge. Same as in WPF.

> That's why DataTemplates property exsits, right?

Yes and no. DataTemplates is lower priority property to define template. Yes, it requires DataType to be set. You can also add a data template from resources.
```xaml
<ListBox>
   <ListBox.DataTemplates>
       <StaticResource ResourceKey="MyDataTemplate1" />
       <StaticResource ResourceKey="MyDataTemplate2" />
   </ListBox.DataTemplates>
</ListBox>
````

Basically, when control needs to find a data template for the content it uses following algorithm:
1. Read ContentTemplate or ItemTemplate if available. DataType needs to be empty or specific to a content type.
2. If not, then reads DataTemplates from the same control. DataType needs to be specific to a content type.
3. If data template still wasn't found, then repeat previous step with parent control (i.e. read DataTemplates collection of the parent). Repeat until matching DataTemplate was found or there is no parent control anymore.

Personally, I would recommend using ContentTemplate or ItemTemplate always when possible (it doesn't accept multiple templates for different types).

> I can creat a DataTemplates in ResourceDictionary

I don't understand why you would put DataTemplates collection in the resources file. It makes a lot of sense for individual data templates which you can later reuse.

**Reply by yll690** [reply]
>  apply to models automatically 

I mean taking effect without reference it explicitly like `ItemTemplate="{StaticResouce MyDataTemplate}"`.

>      <ListBox.DataTemplates>
>       <StaticResource ResourceKey="MyDataTemplate1" />
>       <StaticResource ResourceKey="MyDataTemplate2" />
>     </ListBox.DataTemplates>

I didn't realise `StaticResource` can use like this, thanks for reminding me.

Actually I have a diagram-designer library which contains dozens models and templates. In WPF, the application who uses this library can just add one line of `ResourceDictionary.MergedDictionaries` in App.xaml and then everything is done, They don't need to care which template is corresponding to which model, just need to put models to the interface. Maybe it's a bad practice, but let the library user to explicitly reference every template is also unrealistic, let alone the library user also doesn't know which type of model will be presented.

> I don't understand why you would put DataTemplates collection in the resources file

Or define a `DataTemplates` in a single xaml file. That's doesn't matter. I just want to pack all templates together.

Anyway, the `DataTemplateInclude` meets my needs now. Thank you, I learned lot.
